# Лекция 1. Знакомство со средой. Числа и вывод
Курс: Основы программирования 
Язык: Julia
Среда: JupyterLab

## 0. Зачем инженеру программирование

In [5]:
R = 150        # сопротивление, Ом
I = 0.5        # ток, А

U = I * R      # напряжение, В
P = U * I      # мощность, Вт

println("Напряжение: ", U, " В")
println("Мощность:   ", P, " Вт")

Напряжение: 75.0 В
Мощность:   37.5 Вт


## 1. Среда: JupyterLab

### 1.1 Ядро и блокнот
Блокнот (notebook) — это файл с расширением .ipynb, ваш рабочий документ. Он состоит из ячеек.

Ядро (kernel) — это запущенная в фоне Julia, которая выполняет код из ячеек. Ядро помнит всё, что вы в нём насчитали: если в одной ячейке написать R = 100, то R будет доступна во всех остальных ячейках, пока ядро живо.

Статус ядра виден в правом верхнем углу:

○ пустой кружок — ядро свободно, ждёт;
● закрашенный — ядро занято, ячейка считается.

### 1.2 Два типа ячеек
| Тип | Что внутри | Зачем |
|-----|-----------|-------|
| **Code** | код на Julia | вычисления |
| **Markdown** | форматированный текст | заголовки, пояснения, формулы |

Переключение типа: **Esc**, затем **Y** (в код) или **M** (в текст).

### 1.3 Горячие клавиши

Запуск ячейки (работает в обоих режимах):

| Клавиши | Действие |
|---------|----------|
| **Shift + Enter** | выполнить и перейти к следующей ячейке |
| **Ctrl + Enter** | выполнить и остаться в этой же |
| **Alt + Enter** | выполнить и создать новую ячейку ниже |

В командном режиме (сначала **Esc**):

| Клавиши | Действие |
|---------|----------|
| **A** | вставить ячейку **выше** (above) |
| **B** | вставить ячейку **ниже** (below) |
| **D**, **D** | удалить ячейку (два раза подряд) |
| **Z** | отменить удаление ячейки |
| **M** | сделать ячейку текстовой (markdown) |
| **Y** | сделать ячейку кодовой |
| **I**, **I** | прервать выполнение (interrupt) |
| **0**, **0** | перезапустить ядро (restart) |
| **Ctrl + S** | сохранить блокнот |

### 1.4 Когда всё зависло или запуталось

- **Прервать вычисление** (`I`, `I` или квадратик ■ на панели) — если ячейка считает слишком долго.
- **Restart Kernel** (`0`, `0`) — ядро забывает все переменные и начинает с чистого листа.
- **Restart Kernel and Run All Cells** — перезапустить и прогнать блокнот сверху вниз.

### 1.5 Порядок выполнения

Слева от каждой выполненной ячейки стоит номер: `[1]`, `[2]`, `[3]` — это порядок, в котором ячейки **фактически** запускались. Ядро выполняет ячейки в том порядке, в котором вы нажимаете Shift+Enter, а не сверху вниз.

`[*]` означает, что ячейка выполняется прямо сейчас.

### 1.6 Ошибки

Ошибка — это нормальная часть работы.

In [10]:
sqrt(-1)

LoadError: DomainError with -1.0:
sqrt was called with a negative real argument but will only return a complex result if called with a complex argument. Try sqrt(Complex(x)).

Julia выведет красный текст. Читать в нём нужно **первую строку**: там тип ошибки и её описание. Здесь — `DomainError`: у отрицательного числа нет вещественного квадратного корня.

## 2. Числа и арифметика

### 2.1 Основные операции

In [11]:
7 + 3      # сложение

10

In [12]:
7 - 3      # вычитание

4

In [13]:
7 * 3      # умножение

21

In [14]:
7 / 3      # деление

2.3333333333333335

In [15]:
7 ^ 3      # возведение в степень

343

Если нужна целая часть и остаток:

In [16]:
div(7, 2)   # целая часть от деления: 3

3

In [17]:
rem(7, 2)   # остаток: 1

1

In [18]:
7 % 2       # то же самое, что rem

1

### 2.2 Приоритет операций

Тот же, что в математике: сначала `^`, потом `*` и `/`, потом `+` и `-`.

In [19]:
2 + 3 * 4      # 14, а не 20

14

In [20]:
(2 + 3) * 4    # 20

20

In [21]:
-2 ^ 2         # -4: степень старше минуса

-4

In [22]:
(-2) ^ 2       # 4

4

### 2.3 Дробные числа не точны

In [23]:
0.1 + 0.2

0.30000000000000004

Результат — `0.30000000000000004`. Это не баг Julia. Дробные числа хранятся в двоичной системе с конечной точностью, и `0.1` в ней не представима точно — ровно как `1/3` не записывается точно в десятичной.

Отсюда практическое правило: **дробные числа нельзя сравнивать на точное равенство**. Сравнения разберём в следующей лекции.

## 3. Переменные и константы

### 3.1 Присваивание

In [24]:
a = 5

5

In [25]:
a

5

In [26]:
b = 2.5

2.5

In [27]:
b

2.5

In [28]:
имя_переменной = 100

100

Знак `=` — это не «равно», а «положить значение в переменную». Слева всегда имя, справа — значение или выражение.

In [29]:
x = 10

10

In [30]:
x = x + 1     # взять старое значение, прибавить 1, положить обратно

11

In [31]:
x

11

### 3.2 Правила имён

- разрешены буквы, цифры и `_`, но **начинаться с цифры имя не может**: `2a` — ошибка;
- регистр важен: `R` и `r` — две разные переменные;
- нельзя использовать служебные слова языка (`if`, `for`, `function`, `end` и подобные);
- имя должно говорить, что внутри: `U_ном` лучше, чем `q7`.

### 3.3 Несколько присваиваний

In [32]:
c, d = 3, 4      # c = 3, d = 4

(3, 4)

In [35]:
e1 = e2 = 0      # обе переменные равны нулю

0

### 3.4 Встроенные константы

In [38]:
pi          # 3.141592653589793

π = 3.1415926535897...

In [39]:
Inf         # бесконечность

Inf

In [40]:
-Inf

-Inf

In [41]:
NaN         # «не число» (Not a Number)

NaN

Откуда берутся `Inf` и `NaN`:

In [42]:
1 / 0       

Inf

In [43]:
-1 / 0      

-Inf

In [44]:
0 / 0       

NaN

Julia не падает с ошибкой при делении на ноль в дробных числах — она возвращает `Inf`. Если в ваших расчётах вдруг появился `NaN`, значит где-то раньше произошло `0/0` или `Inf - Inf`.

### 3.5 Небольшая особенность Julia

В Julia имена могут состоять из математических символов. Такой символ набирается как в LaTeX: печатаете `\pi` и нажимаете **Tab**.

In [46]:
π            # \pi + Tab — то же самое, что pi

π = 3.1415926535897...

In [ ]:
√16          # \sqrt + Tab — то же, что sqrt(16)

In [ ]:
7 ÷ 2        # \div + Tab — то же, что div(7, 2)

Выглядит красиво, но набирается медленно.

Между числом и именем переменной знак умножения можно не писать

In [47]:
r = 3

3

In [48]:
2pi * r      # длина окружности, то же самое, что 2 * pi * r

18.84955592153876

## 4. Встроенные функции

Функция вызывается так: имя, круглые скобки, аргументы через запятую.

### 4.1 Общематематические

In [49]:
sqrt(16)      # квадратный корень

4.0

In [50]:
abs(-7)       # модуль

7

In [51]:
exp(2)        # экспонента

7.38905609893065

In [52]:
log(100)      # натуральный логарифм

4.605170185988092

In [53]:
log10(100)    # десятичный

2.0

In [54]:
log2(8)       # двоичный

3.0

In [55]:
log(2, 8)     # логарифм 8 по основанию 2

3.0

### 4.2 Округление

In [57]:
round(3.7)          # 4.0

4.0

In [58]:
floor(3.7)          # 3.0 — вниз

3.0

In [59]:
ceil(3.2)           # 4.0 — вверх

4.0

In [64]:
round(3.14159, digits = 2)   # 3.14 — до двух знаков

3.14

Мелочь, о которой стоит знать:

In [65]:
round(2.5)

2.0

In [66]:
round(3.5)

4.0

При ровной половине Julia округляет к ближайшему **чётному**.

### 4.3 Тригонометрия

Все тригонометрические функции работают в **радианах**.

In [67]:
sin(pi / 2)      # 1.0

1.0

In [69]:
cos(0)           # 1.0

1.0

In [70]:
tan(pi / 4)

0.9999999999999999

In [72]:
deg2rad(90)      # градусы -> радианы

1.5707963267948966

In [71]:
rad2deg(pi)      # радианы -> градусы

180.0

In [74]:
round(sin(deg2rad(30)),digits = 2) # синус 30 градусов

0.5

## 5. Вывод результатов

### 5.1 Автоматический вывод

Блокнот сам показывает значение **последнего выражения** в ячейке. Остальные вычисляются молча.

In [75]:
1 + 1
2 + 2
3 + 3

6

Показалось только `6`.

Точка с запятой в конце подавляет и этот вывод:

In [77]:
a = 3 + 3;

Это удобно, когда результат огромный, а смотреть на него не нужно.

### 5.2 print и println

In [79]:
print("Привет")
print(" мир")

Привет мир

In [80]:
println("Привет")
println("мир")

Привет
мир


`print` печатает как есть. `println` — то же самое плюс переход на новую строку (`ln` = line).

Несколько значений — через запятую:

In [81]:
U = 110
println("Напряжение: ", U, " В")

Напряжение: 110 В


### 5.3 Подстановка значений в текст

Внутри строки можно писать `$имя_переменной` — Julia подставит туда значение.

In [82]:
U = 110
I = 0.5
println("Напряжение $U В, ток $I А")

Напряжение 110 В, ток 0.5 А


Если нужно вычислить прямо в строке — выражение берётся в `$( )`:

In [83]:
println("Мощность: $(U * I) Вт")

Мощность: 55.0 Вт


### 5.4 Служебные символы в строках

| Запись | Что означает |
|--------|--------------|
| `\n` | перевод строки |
| `\t` | табуляция (отступ до следующей позиции) |
| `\\` | сам символ обратной косой черты |
| `\"` | кавычка внутри строки |
| `\$` | сам знак доллара |

In [89]:
println("Первая строка\nВторая строка")

Первая строка
Вторая строка


In [98]:
println("Имя\tЗначение")
println("R\t220")
println("I\t0.5")

Имя	Значение
R	220
I	0.5


In [102]:
println("Путь: C:\\Users\\student")
println("Он сказал: \"работает\"")

Путь: C:\Users\student
Он сказал: "работает"


Обратите внимание: `println` уже добавляет перевод строки в конце, поэтому писать `println("текст\n")` — значит получить лишнюю пустую строку.

### 5.5 display

In [104]:
display(1 / 3)
display("текст")

0.3333333333333333

"текст"

`display` показывает значение в «богатом» виде — так же, как автоматический вывод последней строки. Для чисел разница с `println` невелика, но позже, когда дойдём до таблиц и графиков, `display` будет рисовать их ровными таблицами, а `println` — вываливать текстом.

In [107]:
x = 1 / 3
println(x)
display(x)

0.3333333333333333


0.3333333333333333

Это массив, разберём позже, здесь только как пример объекта с двумя видами вывода.

In [108]:
x = [1 2; 1 2]
println(x)

[1 2; 1 2]


In [109]:
display(x)

2×2 Matrix{Int64}:
 1  2
 1  2

### 5.6 @show — для отладки

In [112]:
y = 42
@show y
@show 2 + 2

42
y = 42
2 + 2 = 4


4

`@show` печатает и само выражение, и его значение. Это самый быстрый способ посмотреть, что происходит внутри вашей программы, когда она ведёт себя не так, как задумано.

## 6. Собираем всё вместе

In [119]:
# Расчёт цепи по закону Ома
U = 220        # напряжение, В
R = 100        # сопротивление, Ом

I = U / R      # ток, А
P = U * I      # мощность, Вт

println("Исходные данные:")
println("\tU = $U В")
println("\tR = $R Ом")
println()
println("Результат:")
println("\tI = $(round(I, digits = 3)) А")
println("\tP = $(round(P, digits = 1)) Вт")

Исходные данные:
	U = 220 В
	R = 100 Ом

Результат:
	I = 2.2 А
	P = 484.0 Вт
